NeoML/src/TraditionalML/Function.cpp

```diff
@@ -17,451 +17,436 @@ limitations under the License.
+struct CThreadFunctionWithHessianImpl::CProductArgs final : public CThreadFunctionWithHessianImpl::IArgs {
+	CProductArgs( CThreadFunctionWithHessianImpl& f, const CFloatVector& argument );
 
-	CArray<CFloatVector> resultReduction;
-	resultReduction.Add( CFloatVector( result.Size() ), curThreadCount );
+	void Run( int threadIndex, int index, int count ) override;
+	void SubRun( int /*index*/, CFloatVectorDesc, double* /*value*/,
+		CFloatVector& /*grad*/, float /*answer*/, float /*weight*/ ) override {}
+	void Reduction() override;
 
-	NEOML_OMP_NUM_THREADS( curThreadCount )
-	{
-		const int threadNumber = OmpGetThreadNum();
-		CFloatVector& privateResult = resultReduction[threadNumber];
-		privateResult.Nullify();
+	const CFloatVector& Argument;
+	CArray<CFloatVector> ResultReduction{};
+	CFloatVector Result{};
+};
 
-		int index = 0;
-		int count = 0;
-		if( OmpGetTaskIndexAndCount( vectorCount, index, count ) ) {
-			for( int i = 0; i < count; i++ ) {
-				if( hessian[index] != 0 ) {
-					CFloatVectorDesc desc;
-					matrix.GetRow( index, desc );
+CThreadFunctionWithHessianImpl::CProductArgs::CProductArgs(
+		CThreadFunctionWithHessianImpl& f, const CFloatVector& argument ) :
+	IArgs( f ),
+	Argument( argument ),
+	Result( argument / F.ErrorWeight )
+{
+	Result.SetAt( Result.Size() - 1, 0 );
+	ResultReduction.Add( CFloatVector( Result.Size() ), ThreadsCount );
+}
 
-					double temp = LinearFunction( arg, desc );
-					temp *= hessian[index];
+void CThreadFunctionWithHessianImpl::CProductArgs::Run( int threadIndex, int index, int count )
+{
+	CFloatVector& result = ResultReduction[threadIndex];
+	result.Nullify();
 
-					privateResult.MultiplyAndAddExt( desc, temp );
-				}
+	for( int i = 0; i < count; ++i, ++index ) {
+		if( F.Hessian[index] != 0 ) {
+			CFloatVectorDesc desc;
+			F.Matrix.GetRow( index, desc );
 
-				index++;
-			}
+			const double temp = LinearFunction( Argument, desc ) * F.Hessian[index];
+			result.MultiplyAndAddExt( desc, temp );
 		}
 	}
+}
 
-	for( int i = 0; i < resultReduction.Size(); i++ ) {
-		result += resultReduction[i];
+void CThreadFunctionWithHessianImpl::CProductArgs::Reduction()
+{
+	for( int i = 0; i < ResultReduction.Size(); ++i ) {
+		Result += ResultReduction[i];
 	}
-
-	return result;
 }
 
 //------------------------------------------------------------------------------------------------------------
 
-CSquaredHinge::CSquaredHinge( const IProblem& data, double _errorWeight, float _l1Coeff, int _threadCount ) :
-	matrix( data.GetMatrix() ),
-	errorWeight( static_cast<float>( _errorWeight ) ),
-	l1Coeff( _l1Coeff ),
-	threadCount( _threadCount ),
-	value( 0.f ),
-	answers( data.GetVectorCount() ),
-	weights( data.GetVectorCount() )
+struct CThreadFunctionWithHessianImpl::CSetArgs : public CThreadFunctionWithHessianImpl::IArgs {
+	CSetArgs( CThreadFunctionWithHessianImpl&, const CFloatVector& argument );
+
+	void Run( int threadIndex, int index, int count ) override;
```

Comment:
`final`? You finalize here `Run` method and allow to override `SubRun`

Reply:
fixed